content/posts/vmware-esx-memory-management.md

```diff
@@ -0,0 +1,46 @@
+
+**Resource sharing**: It is uneconomical for each virtual machine to have all resources available from the hypervisor when it does not need most of them - however, the hypervisor must determine a "fair" way to distribute resources
+
+## Methods
+
+### Memory management
+
+To allow guest OS page eviction policies to function, special drivers called "memory balloon(s)" are installed into them. When the hypervisor needs to reclaim memory, it communicates to the driver to "inflate" the balloon and tricks the guest into thinking more memory is used and then evicts some memory that the hypervisor can then reclaim and allocate to another virtual machine. The balloon stays inflated to ensure the guest retains the specified memory ceiling. When the hypervisor has available resources for the guest to use, the balloon is deflated, allowing memory to be used again. This method generally works, with few exceptions - on Windows, the ballooning driver may not load in quickly enough and memory usage may spike past a set limit, forcing the hypervisor to make non-optimal evictions. Additionally, this relies on guest co-operation, which may not always be available; the hypervisor must have a fallback in case guests are unable to respond to the balloon.
+
+### Data sharing
+
+When multiple guests share the same data (code pages, read-only data), ESX deduplicates the data by adding another layer of indirection from a "physical" (to the operating system) page to a machine (managed by ESX) page. This allows the hypervisor to deduplicate the pages when they are mapped in by hashing the contents of a page and if the hashes match, verifying the content actually matches before remapping the "physical" page. ESX also performs copy-on-write (CoW) so when a guest writes to a shared page, CoW occurs and preserves correctness. However, this becomes untenable at higher-scale due to $O(n^2)$ page search, so ESX picks random times to deduplicate pages. In testing, up to 60% of a guest's memory can be deduplicated, though this is the optimistic case - diverse workloads fared worse due to the time required to check for potential deduplication.
+
+### Resource sharing
+
+ESX uses a ticket/lease mechanism to distribute resources to guests - the more tickets a guest has, the more resources available to them. To prevent a guest from hogging/acquiring all resources and refusing to release them, ESX implements the concept of the "idle memory tax"; the tax comes at reclamation time - guests with more unused resources will be picked first to have resources taken away from them.
+
+## Benchmarks
+
+Many benchmarks were done - refer to the paper for methodology, equations, and results. The readers/author of this summary do not have the pre-requisite background to accurately interpret those results.
```

Comment:
OS MS student saying this... Get PhD asap 😄

Reply:
Yeah I was being lazy 😅 - I'll read the benchmark section of the paper and revise this PR

Reply:
Don't stress it. imo I think this captures the concepts of the paper well.